go/vt/vtgate/vindexes/vschema_routing_test.go

```diff
@@ -0,0 +1,481 @@
+					_, ok := allTables[tname]
+					if !ok {
+						allTables[tname] = true
+					}
+				}
+			}
+			for _, ksName := range tc.explicit {
+				source.Keyspaces[ksName].RequireExplicitRouting = true
+			}
+			vschema := BuildVSchema(source, sqlparser.NewTestParser())
+			require.NotNil(t, vschema)
+			AddAdditionalGlobalTables(source, vschema)
+
+			var globalTables, ambiguousTables []string
+			for tname := range allTables {
+				isGlobal, isAmbiguous := isTableGloballyRoutable(vschema, tname)
+				if isGlobal {
+					globalTables = append(globalTables, tname)
+				} else if isAmbiguous {
+					ambiguousTables = append(ambiguousTables, tname)
+				}
+			}
+			sort.Strings(globalTables)
+			sort.Strings(ambiguousTables)
+			sort.Strings(tc.expGlobalTables)
+			sort.Strings(tc.expAmbiguousTables)
+			require.EqualValuesf(t, tc.expGlobalTables, globalTables, "global tables mismatch")
+			require.EqualValuesf(t, tc.expAmbiguousTables, ambiguousTables, "ambiguous tables mismatch")
+		})
+	}
+}
+
+// TestAutoGlobalRouting tests adding tables in unsharded keyspaces to global routing if they don't have
+// an associated VSchema which has the RequireExplicitRouting flag set. These tables should also not be
+// already part of the global routing tables via the VSchema of sharded keyspaces.
+func TestAutoGlobalRoutingBasic(t *testing.T) {
+	// Create two unsharded keyspaces and two sharded keyspaces, each with some common tables.
+	unsharded1 := &vschemapb.Keyspace{
+		Tables: map[string]*vschemapb.Table{
+			"table1":   {}, // unique, should be added to global routing
+			"table2":   {}, // unique, should be added to global routing
+			"scommon1": {}, // common with sharded1, should not be added to global routing because it is already global because of sharded1
+			"ucommon3": {}, // common with unsharded2, should not be added to global routing because it is ambiguous because of unsharded2
+		},
+	}
+	unsharded2 := &vschemapb.Keyspace{
+		Tables: map[string]*vschemapb.Table{
+			"table3":   {}, // unique, should be added to global routing
+			"table4":   {}, // unique, should be added to global routing
+			"scommon1": {}, // common with sharded1, should not be added to global routing because it is already global because of sharded1
+			"scommon2": {}, // common with sharded1, should not be added to global routing because it is already global because of sharded1
+			"ucommon3": {}, // common with unsharded1, should not be added to global routing because it is ambiguous because of unsharded1
+		},
+	}
+	sharded1 := &vschemapb.Keyspace{
+		Sharded: true,
+		Vindexes: map[string]*vschemapb.Vindex{
+			"xxhash": {
+				Type: "xxhash",
+			},
+		},
+		Tables: map[string]*vschemapb.Table{
+			"table5":   {}, // unique, should be added to global routing
+			"scommon1": {}, // common with unsharded1 and unsharded2, should be added to global routing because it's in the vschema
+			"scommon2": {}, // common with unsharded2, ambiguous because of sharded2
```

Comment:
The comment seems wrong. It is not ambiguous because of sharded2 because sharded2 requires explicit routing. We even assert that `scommon2` can be globally routed.